[typescript] Add TS typings for v1.0.0-beta.2

sebald opened the pull request:
<!-- Thanks so much for your PR, your contribution is appreciated! -->

As discussed in #7522 here is an initial set of TypeScript typings. The typings include the full public API and all components provided by `material-ui` (v1.0.0-beta.2). There are not typings for `src/internal/*` yet, except the `Transition` component.

If you guys want to go down this road and bundle typings with `material-ui` the next step would be to add tests for the typings. Where tests are just some sample `.ts` files that with sample code. You can find an example inside the [`redux` repo](https://github.com/reactjs/redux/blob/master/test/typescript/store.ts).

Let me know if you want tests immediately.

For comparison these are the [typings for the `0.x` branch]((https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/material-ui/index.d.ts)).

---

- [ ] PR has tests / docs demo, and is linted.
- [x] Commit and PR titles begin with [ComponentName], and are in imperative form: "[Component] Fix leaky abstraction".
- [x] Description explains the issue / use-case resolved, and auto-closes the related issue(s) (http://tr.im/vFqem).

oliviertassinari commented:
@sebald Thanks for the PR! 

> the next step would be to add tests for the typings.

That sounds like a good idea. I think that we can move forward with this PR and add them later.
From what I can read of the generated types, we gonna definitely need a tool to generate them.
For instance, the `propTypes`API will change PR after PR. We can't expect/ask for people to keep that up to date.

alitaheri commented on package.json:
```diff
@@ -72,6 +73,8 @@
   },
   "devDependencies": {
+    "@types/enzyme": "^2.8.4",
+    "@types/react": "^15.0.39",
```
These shouldn't be dev dependencies. They are needed for the typings to function.

alitaheri commented on typings/index.d.ts:
```diff
@@ -0,0 +1,1909 @@
+/*                                               */
+/* ============================================= */
+
+declare module 'material-ui/AppBar' {
```
Why not make individual files matching their components? I think that will make it easier to maintain. Then you could require them here.

sebald replied:
And re-create the `src` folder structure inside of the `typings` folder?
Guess it would also be possible to include `index.d.ts` files in the corresponding component folder. This would make the `typings` property in the `package.json` redudant.

I chose to declare modules inside a single file to "adhere" what was done with the `0.x` typings :)

oliviertassinari replied:
Still, I think that the file should be moved to the src folder. We use a `copy-files.js` script to flatten the folder structure before publishing.

sebald replied:
Sure, going to move the files today and update the script accordingly.

oliviertassinari replied:
Might not need to update the script.

sebald replied:
Yeah, I meant that I will test, if the script is picking the `.d.ts` file up and copying it :)

sebald replied:
I started migrating from one file to dedicated files inside the source folder. One issue arose with that:

For every `js` file there has to be a `.d.ts` file, which would lead to folders structures like the following:

![screen shot 2017-07-27 at 13 34 17](https://user-images.githubusercontent.com/985701/28668487-53578bc0-72d0-11e7-9e82-bbf8d69876e0.png)

Are you sure you really want that? This might be annoying for all contributors that do not care about the typings.

@alitaheri Do you know a way to pack all typings inside a single `index.d.ts` and have alias imports? So that I do not have to create a separate file to support imports like `material-ui/AppBar/AppBar`?

oliviertassinari replied:
Maybe splitting the typescript file is too much work to begin with?

alitaheri replied:
I think it's too much work too 😅 

Leave it like this. We might slowly move the files around later on. Or build a tool that generates the typings anyway. Don't waste too much time on it.

sebald replied:
I made some fixes to the single file typings, but didn't had the time to push them here. Will push first thing on Monday. ☺️

sebald replied:
... And also resolve the new conflicts.

oliviertassinari replied:
Sounds good. Let's merge it on monday if everybody agree.

alitaheri commented on typings/index.d.ts:
```diff
@@ -0,0 +1,1909 @@
+///<reference types="react" />
+///<reference types="enzyme" />
+
+declare namespace MaterialUI {
```
namespaces aren't easy to work with. I suggest using es6 module system. Typescript is good at understanding exports.

sebald replied:
That's actually a common practice. Otherwise you would have to prefix everything with `Mui` or something.

sebald replied:
The reason why I used a namespace is b/c I didn't want to import the specialized component I created for JSS-styled components, like:

```ts
export default class Button extends MaterialUI.Component<ButtinProps> {}
```

Every module would have to import that declaration otherwise.

alitaheri commented on typings/index.d.ts:
```diff
@@ -0,0 +1,1909 @@
+declare module 'material-ui/Button/Button' {
+  export interface ButtonProps {
+    color?: MaterialUI.PropTypes.Color | 'contrast';
+    component?: React.ReactNode;
```
This isn't a node, it's a component ( string e.g. 'div' or function ) **not** element or node.

Change it to `React.SFC | React.ComponentClass | string` ( I think )

alitaheri commented on typings/index.d.ts:
```diff
@@ -0,0 +1,1909 @@
+
+  export interface MuiThemeProviderProps {
+    theme?: Theme<any>;
+    sheetsManager?: Object;
```
`Object`?

sebald replied:
That's what the `PropTypes` say :)
https://github.com/callemall/material-ui/blob/73a6d33072eabd93d1c78e9ae1be781d21601c7c/src/styles/MuiThemeProvider.js#L77-L92

alitaheri commented on typings/index.d.ts:
```diff
@@ -0,0 +1,1909 @@
+
+declare module 'material-ui/styles/colorManipulator' {
+  export type ColorFormat = 'rgb' | 'rgba' | 'hsl' | 'hsla';
+  export type ColorObject = {
```
This could be a disjoint type:

```ts
export type ColorObject = 
  | { type: 'rgb' | 'hsl'; color: [number, number, number] }
  | { type: 'rgba' | 'hsla'; color: [number, number, number, number] }
```

alitaheri commented on typings/index.d.ts:
```diff
@@ -0,0 +1,1909 @@
+   */
+  export default function createGenerateClassName(): (
+    rule: Object,
+    stylesheet?: Object
```
Don't use `Object`. either use `object` or `any`

sebald commented:
> From what I can read of the generated types, we gonna definitely need a tool to generate them.
For instance, the propTypesAPI will change PR after PR. We can't expect nor ask for people to keep that up to date.

@oliviertassinari If this works well, this is definitely the better approach. I wonder how good tools like `tsgen` are and if they can create good typings. But currently there are a still a lot of `PropTypes` AFAIK. (Is there no codemod/tool to create flow types from them?)

Also, if there are flow types, it's mostly copy/pasting them into an `.d.ts` file. HOCs need some attention and some type can be stronger (I guess?). For example for a property like `component?: string | Function` the TS typings for React already have some fixed types.

thupi commented:
Hey, i just tried tis typings and the quality of the typing seams very good and works great. However, i am experiencing some issue when it comes to native element props such as onClick on button etc.

Am i doing something wrong :-) ?

sercand commented:
@thupi you are not alone. Also, there are some places where even documentation code does not work either.

sercand commented:
@sebald I have modified "material-ui/scripts/build-api-docs.js" script to generate typings for the components for an internal company project. It worked very well. I think adding a script to create props should be considered.

oliviertassinari commented:
@sercand That sounds great. Do you mind sharing that script and the documentation issue you have noticed? :)

sebald commented:
@thupi @sercand No, that's completely my fault. I accidentally removed `ButtonBase` from the typings, which is an internal helper that has adds the handlers :-x

I already fixed that, but didn't had time on Friday to push the update here. As said, will do tomorrow :)

Would also love to see the script! Just one caveat I have with scripts generating typings: the `PropTypes` and flow types often define `Function` for handlers in MUI, which is totally fine, but the typings won't tell you that an event will be passed to the handler callback. I replaced that with the corresponding `React.ReactEventHandler<T>` in order to have more helpful typings :)

thupi commented:
@sebald Ahh, okay that makes sense :-) ! But i just wanne say the the typings realy is very helpful and only needs some minor tweaks in order to work perfectly :-) 

I will continue using the typings in order to test i further :-) Thanks a lot for the great progress :-) !

sebald commented:
@oliviertassinari Pushed the aforementioned fixes. As far as I'm concerned the typings are in an healthy spot and can be merged.

As said, next step is would be to add a test suite to prevent regressions :-)
Where tests is actually just ah `.ts.` file using the typings and to check if TS can parse it...

If anyone runs into troubles/has issues with the typings, please let my know! Will fix as soon as possible. I also will track the releases to see if there are changes in the API.

sercand commented:
@sebald The props type of [ListItemText](https://github.com/PTW-Freiburg/material-ui/blob/ts-typings-1.0.0-beta.2/typings/index.d.ts#L853) is wrong it must be ListItemTextProps.

sebald commented:
@sercand Thx, fixed :)

---

Currently typings for the Icons are missing. We're using [a script](https://gist.github.com/sebald/6405bce76912a7e504b8569afdb9ca6c) to generate them. Maybe this should be added to `packages/material-ui-icons`.

oliviertassinari commented:
@sebald Thanks a lot!

thupi commented:
@sebald The Drawer component has theme as a required prop. That seems not to be the case in the docs :-)

sebald commented:
@thupi Thanks! See #7606

@oliviertassinari I am not sure how you want me to proceed further. So, I opened a new PR, but I think it is not necessary to make a PR for every small fix to the typings. I will use #7606 to gather all the issues people are running into. 

Maybe you could merge that updates when you'll release the next beta?

oliviertassinari commented:
@sebald Sounds good to me 👍

stevegeek commented on typings/index.d.ts:
```diff
@@ -0,0 +1,2022 @@
+    onRequestClose?: React.EventHandler<any>;
+    open?: boolean;
+    SlideProps?: Object;
+    theme: Theme;
```
Thanks for the typings! This doesn't look right though, theme param and its not optional?

stevegeek replied:
Ah just saw the other PR

newash commented:
hey @sebald 

I must be doing something wrong here, but I have tons of "Duplicate identifier" errors for this file.
Like for `AppBar` here:
```typescript
export { default as AppBar } from 'material-ui/AppBar';
```
And VSCode brings me to this like as the definition of `AppBar`:
```typescript
export default class AppBar extends MaterialUI.Component<AppBarProps> {}
```

What is your TS setup? Versions? Any special settings?

sebald commented:
Guess you should open a new issue :) + please create a repo that reproduces the error.
I don't have any special settings. I am using TS 2.4.2.
